src/toast/ops/sim_gaindrifts.py

```diff
@@ -0,0 +1,322 @@
+        log = Logger.get()
+
+        for ob in data.obs:
+            # Get the detectors we are using for this observation
+            dets = ob.select_local_detectors(detectors)
+            if len(dets) == 0:
+                # Nothing to do for this observation
+                continue
+            comm = ob.comm
+            rank = ob.comm_rank
+            # Make sure detector data output exists
+            ob.detdata.ensure(self.det_data, detectors=dets)
+            obsindx = ob.uid
+            telescope = ob.telescope.uid
+            focalplane = ob.telescope.focalplane
+            size = ob.detdata[self.det_data][dets[0]].size
+            fsampl = focalplane.sample_rate.to_value(u.Hz)
+
+            if self.drift_mode == "linear_drift":
+                key1 = (
+                    self.realization * 4294967296 + telescope * 65536 + self.component
+                )
+                counter1 = 0
+                counter2 = 0
+
+                for det in dets:
+                    detindx = focalplane[det]["uid"]
+                    key2 = obsindx * 4294967296 + detindx
+
+                    rngdata = rng.random(
+                        1,
+                        sampler="gaussian",
+                        key=(key1, key2),
+                        counter=(counter1, counter2),
+                    )
+                    gf = 1 + rngdata[0] * self.sigma_drift
+                    gain = (gf - 1) * np.linspace(0, 1, size) + 1
+
+                    ob.detdata[self.det_data][det] *= gain
+
+            elif self.drift_mode == "thermal_drift":
+                fmin = fsampl / (4 * size)
+                # the factor of 4x the length of the sample vector  is
+                # to avoid circular correlations
+                freq = np.logspace(np.log10(fmin), np.log10(fsampl / 2.0), 1000)
+                psd = self.get_psd(freq)
+                det_group = np.unique(focalplane.detector_data[self.focalplane_group])
+                thermal_fluct = []
+                for iw, w in enumerate(det_group):
+                    # simulate a noise-like timestream
+                    gain = sim_noise_timestream(
+                        realization=self.realization,
+                        telescope=ob.telescope.uid,
+                        component=self.component,
+                        obsindx=ob.uid,
+                        # we generate the same timestream for the
+                        # detectors in the same group
+                        detindx=iw,
+                        rate=fsampl,
+                        firstsamp=ob.local_index_offset,
+                        samples=ob.n_local_samples,
+                        freq=freq,
+                        psd=psd,
+                        py=False,
+                    )
+                    thermal_fluct.append(gain)
+                thermal_fluct = np.array(thermal_fluct)
+
+                for det in dets:
+                    detindx = focalplane[det]["uid"]
```

Comment:
Note that this detector unique ID is not necessarily a simple index.  It may be a very large integer generated from a hash of the detector name:  https://github.com/hpc4cmb/toast/blob/57f57d12cce917da9d87a18d78e1919167bb56af/src/toast/instrument.py#L322
Something to keep in mind when using it for RNG state.